module/zfs/dsl_dataset.c

```diff
@@ -2353,6 +2353,10 @@ dsl_dataset_promote_check(void *arg, dmu_tx_t *tx)
 	}
 
 	snap = list_head(&ddpa->shared_snaps);
+	if (snap == NULL) {
+		err = SET_ERROR(EXDEV);
```

Comment:
The fix itself looks good but I'd suggest returning `ENOENT` in this case and the next one.  This will be propagated back through the zfs ioctl to the `zfs promote` command and converted in to a more meaningful `dataset does not exist` error message.  Using `EXDEV` should generate a message along the lines of `cross-device link` which isn't as clear.  See the `zfs_standard_error_fmt()` function for the error messages.  Using a different error will also let developers distinguish between this error case and the DS_FLAG_NOPROMOTE case.

```